avm/src/vm/opcode.rs

```diff
@@ -2349,14 +2350,27 @@ pub static OPCODES: Lazy<HashMap<i64, ByteOpcode>> = Lazy::new(|| {
       let hand_mem_ref = Arc::new(hand_mem);
       let fractal = hand_mem_ref.read_fractal(args[0]);
       let subhandler = HandlerFragment::new(args[1], 0);
-      let mut runners = Vec::with_capacity(fractal.len());
-      for i in 0..fractal.len() {
+      let n = num_cpus::get();
+      let l = fractal.len();
+      let s = l / n;
+      let mut runners = Vec::with_capacity(s);
+      for _ in 0..s {
+        runners.push(FuturesUnordered::<HMFuture>::new());
+      }
+      for i in 0..l {
         let mut hm = HandlerMemory::fork(Arc::clone(&hand_mem_ref));
         hm.register_out(args[0], i, CLOSURE_ARG_MEM_START + 1);
         hm.write_fixed(CLOSURE_ARG_MEM_START + 2, i as i64);
-        runners.push(subhandler.clone().run(hm));
+        let j = i / s;
+        runners[j].push(Box::pin(subhandler.clone().run_local(hm)));
       }
-      let mut hms = join_all(runners).await;
+      let mut hms: Vec<HandlerMemory> = Vec::with_capacity(l);
+      for runner in runners {
+        let mut chuncked_hms = task::spawn(async move { runner.collect().await })
```

Comment:
this code completes the parallel version of https://github.com/ledwards/advent-2020/blob/main/day01b.ln in:

```
$ time ./avm/target/release/alan run ../out1.agc 

real	1m52.170s
user	4m39.735s
sys	2m18.951s
```
However, when we try to call `join_all` on the tasks  instead of awaiting, it takes forever without the OS killing the process in my laptop

Reply:
Perhaps it has to do with the fact that the opcodes are all cpu bound so the `block_in_place` call is not getting properly parallelized? Here is the alternate code calling `join_all` on the tokio `task` https://gist.github.com/depombo/9f2d8bfb53587329fad0a7db28978876